File: src/parser/parser.go
package parser

import (
	"fmt"

	"github.com/DomioKing653/FlowScript/src/ast"

	"github.com/DomioKing653/FlowScript/src/lexer"
)

type parser struct {
	tokens []lexer.Token
	pos    int
}

func Parse(tokens []lexer.Token) ast.BlockStatment {
	Body := make([]ast.Statement, 0)
	p := CreateParser(tokens)
	for p.hasTokens() {
		Body = append(Body, parse_stmt(p))
	}
	return ast.BlockStatment{Body: Body}
}

// HELPER METHODS

func CreateParser(tokens []lexer.Token) *parser {
	CreateLookups()
	return &parser{
		tokens: tokens, pos: 0,
	}
}

func (p *parser) currentToken() lexer.Token {
	return p.tokens[p.pos]
}

func (p *parser) currentTokenKind() lexer.TokenKind {
	return p.currentToken().Kind
}

func (p *parser) advance() lexer.Token {
	tk := p.currentToken()
	p.pos++
	return tk
}

func (p *parser) hasTokens() bool {
	return p.pos < len(p.tokens) && p.currentTokenKind() != lexer.EOF
}

func (p *parser) expectError(expectedKind lexer.TokenKind, err any) lexer.Token {
	token := p.currentToken()
	kind := token.Kind

	if kind != expectedKind {
		if err == nil {
			err = fmt.Sprintf("Error::Syntax->Expected %s but received %s", lexer.TokenKindString(expectedKind), lexer.TokenKindString(kind))
		}

		panic(err)
	}

	return p.advance()
}

func (p *parser) expect(expectedKind lexer.TokenKind) lexer.Token {
	return p.expectError(expectedKind, nil)
}

File: src/parser/parser.zig
const stmt = @import("../ast/statement.zig");
const parseStmt = @import("parseStmt.zig");
const Tokens = @import("../lexer/tokens.zig");
const std = @import("std");
pub const Parser = struct {
    tokens: []Tokens.Token,
    pos_idx: usize,
    current_token: Tokens.Token,
    statements: std.ArrayList(stmt.Statement),
    alloc: std.mem.Allocator,
    pub fn parse(self: *Parser) !void {
        self.current_token = self.tokens[0];
        self.pos_idx += 1;
        while (self.current_token.Kind != Tokens.TokenKind.EOF) {
            try self.statements.append(self.alloc, try parseStmt.parseStmt(self));
        }
        std.debug.print("Parsing ended", .{});
    }
    pub fn advance(self: *Parser) !void {
        self.current_token = self.tokens[self.pos_idx];
        self.pos_idx += 1;
    }

    pub fn expect(self: *Parser, kind: Tokens.TokenKind) ![]u8 {
        if (self.current_token.Kind != kind) {
            return parseStmt.ParserErrors.UnexpectedToken;
        } else {
            return self.alloc.dupe(u8, self.current_token.Value);
        }
    }
};
